apps/OboeTester/app/src/main/java/com/mobileer/oboetester/RoundTripLatencyActivity.java

```diff
@@ -165,30 +201,287 @@ public void start() {
+                        nextBursts = mMiddleBufferBursts;
+                        mState = STATE_MEASURE_MIDDLE;
+                    } else {
+                        result = RESULT_OK;
+                        mMessage = "DSP position looks good";
+                        mState = STATE_DONE;
+                    }
+                    break;
+                case STATE_MEASURE_MIDDLE:
+                    mMiddleBufferLatency = latencyFrames;
+                    mMiddleBufferBursts = actualBufferBursts;
+                    // Check to see which side is bad.
+                    if (confidence < 0.5) {
+                        // We may have landed on the DSP so we got a scrambled result.
+                        result = RESULT_DISCONTINUITY;
+                        mMessage = "on top of DSP!";
+                        mState = STATE_DONE;
+                    } else  {
+                        double deviationLow = measureLatencyLinearity(
+                                mLowBufferBursts,
+                                mLowBufferLatency,
+                                mMiddleBufferBursts,
+                                mMiddleBufferLatency);
+                        double deviationHigh = measureLatencyLinearity(
+                                mMiddleBufferBursts,
+                                mMiddleBufferLatency,
+                                mHighBufferBursts,
+                                mHighBufferLatency);
+                        if (deviationLow >= deviationHigh) {
+                            // bottom half was bad so subdivide it
+                            mHighBufferBursts = mMiddleBufferBursts;
+                            mHighBufferLatency = mMiddleBufferLatency;
+                            mMessage = "low half not linear";
+                        } else {
+                            // top half was bad so subdivide it
+                            mLowBufferBursts = mMiddleBufferBursts;
+                            mLowBufferLatency = mMiddleBufferLatency;
+                            mMessage = "high half not linear";
+                        }
+                    }
+                    break;
+                default:
+                    break;
+            }
+
+            if (result == RESULT_UNDEFINED) {
+                if (nextBursts < 0) {
+                    if ((mHighBufferBursts - mLowBufferBursts) <= 1) {
+                        result = RESULT_DISCONTINUITY;
+                        mMessage = "ERROR - DSP position error between "
+                                + mLowBufferBursts + " and "
+                                + mHighBufferBursts + " bursts!";
+                    } else {
+                        // Subdivide the remaining search space.
+                        mMiddleBufferBursts = (mHighBufferBursts + mLowBufferBursts) / 2;
+                        result = mMiddleBufferBursts;
```

Comment:
I'm a bit confused about what we are returning here. Can you add some comments/rename some variables?